### PR TITLE
[FLINK-6968][Table API & SQL] Add Queryable table sink.

### PR DESCRIPTION
**Pull request by liurenjie1024**
## What is the purpose of the change

Streaming tables with unique key are continuously updated. For example queries with a non-windowed aggregation generate such tables. Commonly, such updating tables are emitted via an upsert table sink to an external datastore (k-v store, database) to make it accessible to applications.

## Brief change log

  - *Add a QueryableStateTableSink.*
  - *States are queryable.*

## Verifying this change

This change added tests and can be verified as follows:
  - *Added test that validates that states will be stored.*
  - *Manually verified the change by running a 4 node cluser with 2 JobManagers and 4 TaskManagers, a stateful streaming program, and use QueryableStateClient to test that.*

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? docs

### CODE REVIEW

**Comment by liurenjie1024** [comment]
@twalthr Please help to review this.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sinks/QueryableTableSink.scala [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+    val processFunction = new QueryableStateProcessFunction(
+      namePrefix,
+      queryConfig,
+      keys,
+      getFieldNames,
+      getFieldTypes)
+
+    dataStream.keyBy(new RowKeySelector(keyIndices, keySelectorType))
+      .process(processFunction)
```
Comment (xccui) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sinks/QueryableTableSink.scala]: This `process(processFunction)` method has been deprecated. Replace it with `process(KeyedProcessFunction)`.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sinks/QueryableTableSink.scala [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+  }
+
+  /**
+    * Specifies whether the [[Table]] to write is append-only or not.
+    *
+    * @param isAppendOnly true if the table is append-only, false otherwise.
+    */
+  override def setIsAppendOnly(isAppendOnly: JBool): Unit = {
+    if (isAppendOnly) {
+      throw new IllegalArgumentException("A QueryableTableSink can not be used with append-only tables " +
```
Comment (xccui) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sinks/QueryableTableSink.scala]: This line is too long (should be less than 100 characters).

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sinks/QueryableTableSink.scala [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+import org.slf4j.LoggerFactory
+
+class QueryableTableSink(private val namePrefix: String,
+                         private val queryConfig: StreamQueryConfig)
+  extends UpsertStreamTableSink[Row]
+  with TableSinkBase[JTuple2[JBool, Row]] {
+  private var keys: Array[String] = _
+
+  /**
+    * Configures the unique key fields of the [[Table]] to write.
```
Comment (xccui) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sinks/QueryableTableSink.scala]: Use the full class name for `[[Table]]` since it's not imported.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sinks/QueryableTableSink.scala [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+import org.apache.flink.streaming.api.datastream.DataStream
+import org.apache.flink.streaming.api.functions.ProcessFunction
+import org.apache.flink.table.api.StreamQueryConfig
+import org.apache.flink.table.runtime.aggregate.ProcessFunctionWithCleanupState
+import org.apache.flink.types.Row
+import org.apache.flink.util.Collector
+import org.slf4j.LoggerFactory
+
+class QueryableTableSink(private val namePrefix: String,
+                         private val queryConfig: StreamQueryConfig)
```
Comment (xccui) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sinks/QueryableTableSink.scala]: Format the code like that.
```
class QueryableTableSink(
    private val namePrefix: String,
    private val queryConfig: StreamQueryConfig)
  extends UpsertStreamTableSink[Row]
...
```

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sinks/QueryableTableSink.scala [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+import org.apache.flink.api.java.tuple.{Tuple2 => JTuple2}
+import org.apache.flink.api.java.typeutils.{ResultTypeQueryable, RowTypeInfo}
+import org.apache.flink.configuration.Configuration
+import org.apache.flink.streaming.api.datastream.DataStream
+import org.apache.flink.streaming.api.functions.ProcessFunction
+import org.apache.flink.table.api.StreamQueryConfig
+import org.apache.flink.table.runtime.aggregate.ProcessFunctionWithCleanupState
+import org.apache.flink.types.Row
+import org.apache.flink.util.Collector
+import org.slf4j.LoggerFactory
```
Comment (xccui) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sinks/QueryableTableSink.scala]: Remove unused imports.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sinks/QueryableTableSink.scala [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+      keys,
+      getFieldNames,
+      getFieldTypes)
+
+    dataStream.keyBy(new RowKeySelector(keyIndices, keySelectorType))
+      .process(processFunction)
+  }
+
+  /** Return a deep copy of the [[TableSink]]. */
+  override protected def copy: TableSinkBase[JTuple2[JBool, Row]] = {
```
Comment (xccui) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sinks/QueryableTableSink.scala]: The docs for overridden methods could be omitted.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/stream/table/QueryableTableSinkTest.scala [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    val keys = Array("id")
+    val keyType = new RowTypeInfo(TypeInformation.of(classOf[String]))
+    val fieldNames = Array("id", "is_manager", "name")
+    val fieldTypes: Array[TypeInformation[_]] = Array(
+      TypeInformation.of(classOf[String]).asInstanceOf[TypeInformation[_]],
+      TypeInformation.of(classOf[JBool]).asInstanceOf[TypeInformation[_]],
+      TypeInformation.of(classOf[String]).asInstanceOf[TypeInformation[_]])
+    val func = new QueryableStateProcessFunction("test", queryConfig, keys, fieldNames, fieldTypes)
+
+    val operator = new LegacyKeyedProcessOperator[Row, JTuple2[JBool, Row], Void](func)
```
Comment (xccui) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/stream/table/QueryableTableSinkTest.scala]: Try to avoid using deprecated classes/methods.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/stream/table/QueryableTableSinkTest.scala [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+
+    val testHarness = createHarnessTester(operator,
+      new RowKeySelector(Array(0), keyType),
+      keyType)
+
+    testHarness.open()
+
+
```
Comment (xccui) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/stream/table/QueryableTableSinkTest.scala]: Remove extra blank lines.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sinks/QueryableTableSink.scala (hunk 1 above)
Comment (liurenjie1024) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sinks/QueryableTableSink.scala]: I'm waiting another [PR](https://github.com/apache/flink/pull/5680) to be merged.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/stream/table/QueryableTableSinkTest.scala (hunk 7 above)
Comment (liurenjie1024) [reply flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/stream/table/QueryableTableSinkTest.scala]: Same as above.

**Comment by liurenjie1024** [comment]
@xccui Thanks for you suggestions.

1. Add an it test for this is difficult. Think about it, I need to ensure that elements to be processed while keeping the job running, and it's difficult to achieve this. But I will do some manual test for that.
2. I'll add doc for that.
3. I'll squash the commits when review is done.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sinks/QueryableTableSink.scala [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+import org.apache.flink.streaming.api.datastream.DataStream
+import org.apache.flink.streaming.api.functions.ProcessFunction
+import org.apache.flink.table.api.StreamQueryConfig
+import org.apache.flink.table.runtime.aggregate.ProcessFunctionWithCleanupState
+import org.apache.flink.types.Row
+import org.apache.flink.util.Collector
+import org.slf4j.LoggerFactory
+
+class QueryableTableSink(
+  private val namePrefix: String,
```
Comment (xccui) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sinks/QueryableTableSink.scala]: Four spaces indent for parameter declaring (which is also applicable for methods).

Comment (liurenjie1024) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sinks/QueryableTableSink.scala]: Fixed.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sinks/QueryableTableSink.scala [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+  override def processElement(
+    value: JTuple2[JBool, Row],
+    ctx: ProcessFunction[JTuple2[JBool, Row], Void]#Context,
+    out: Collector[Void]): Unit = {
+    if (value.f0) {
+      for (i <- nonKeyIndices.indices) {
+        states(i).update(value.f1.getField(nonKeyIndices(i)))
+      }
+
+      val currentTime = ctx.timerService().currentProcessingTime()
```
Comment (walterddr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sinks/QueryableTableSink.scala]: Seems like this is registered to work with processing time domain only. I was wondering would it be easier to do ITCase if the table sink also works with event time, or having another piece of event time queryable table sink ?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sinks/QueryableTableSink.scala [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+import org.apache.flink.configuration.Configuration
+import org.apache.flink.streaming.api.datastream.DataStream
+import org.apache.flink.streaming.api.functions.ProcessFunction
+import org.apache.flink.table.api.StreamQueryConfig
+import org.apache.flink.table.runtime.aggregate.ProcessFunctionWithCleanupState
+import org.apache.flink.types.Row
+import org.apache.flink.util.Collector
+import org.slf4j.LoggerFactory
+
+class QueryableTableSink(
```
[the remark this answers is not in the document]
Comment (liurenjie1024) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sinks/QueryableTableSink.scala]: Java doc added.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sinks/QueryableTableSink.scala (hunk 10 above)
Comment (liurenjie1024) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sinks/QueryableTableSink.scala]: I think it's right to be able to process both time domain, and I've added.

**Comment by suez1224** [comment]
Could you please rebase your pr to resolve conflict? Thanks.

**Comment by liurenjie1024** [comment]
@suez1224 Conflict resolved.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sinks/QueryableTableSink.scala [hunk 12]
```diff
@@ -0,0 +1,223 @@
...
+    value: JTuple2[JBool, Row],
+    ctx: KeyedProcessFunction[Row, JTuple2[JBool, Row], Void]#Context,
+    out: Collector[Void]): Unit = {
+    if (value.f0) {
+      for (i <- nonKeyIndices.indices) {
+        states(i).update(value.f1.getField(nonKeyIndices(i)))
+      }
+
+      val currentTime: Long = cleanupTimeDomain match{
+        case TimeDomain.EVENT_TIME => ctx.timestamp()
```
Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sinks/QueryableTableSink.scala]: By default this timestamp is empty. It is filled by a preceding operator that is generated by the Table API.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sinks/QueryableTableSink.scala [hunk 13]
```diff
@@ -0,0 +1,223 @@
...
+
+    initCleanupTimeState("QueryableStateCleanupTime")
+  }
+
+  override def processElement(
+    value: JTuple2[JBool, Row],
+    ctx: KeyedProcessFunction[Row, JTuple2[JBool, Row], Void]#Context,
+    out: Collector[Void]): Unit = {
+    if (value.f0) {
+      for (i <- nonKeyIndices.indices) {
```
Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sinks/QueryableTableSink.scala]: We don't use Scala magic in runtime functions for performance reasons. Use a basic while loop instead.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/util/KeyedOneInputStreamOperatorTestHarness.java [hunk 14]
```diff
@@ -78,6 +82,20 @@ public int numKeyedStateEntries() {
 		}
 	}
 
+	public <S extends State> S getState(K key, StateDescriptor<S, ?> stateDesc) throws Exception {
```
Comment (twalthr) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/util/KeyedOneInputStreamOperatorTestHarness.java]: Is this change necessary? We should only modify code outside of `flink-table` if it is urgently needed.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/stream/table/QueryableTableSinkTest.scala [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+    var expectedData = Array(
+      Row.of(JBool.valueOf(true), "jeff"),
+      Row.of(JBool.valueOf(false), "dean"))
+    var storedData = Array(
+      Row.of(stateOf(key1, stateDesc1), stateOf(key1, stateDesc2)),
+      Row.of(stateOf(key2, stateDesc1), stateOf(key2, stateDesc2)))
+
+    verify(expectedData, storedData)
+
+
```
Comment (twalthr) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/stream/table/QueryableTableSinkTest.scala]: Add comments to this test and remove empty lines here.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/KeyedProcessFunctionWithCleanupState.scala [hunk 16]
```diff
@@ -44,8 +45,20 @@ abstract class KeyedProcessFunctionWithCleanupState[K, I, O](queryConfig: Stream
   protected def registerProcessingCleanupTimer(
     ctx: KeyedProcessFunction[K, I, O]#Context,
     currentTime: Long): Unit = {
-    if (stateCleaningEnabled) {
+    registerCleanupTimer(ctx, currentTime, TimeDomain.PROCESSING_TIME)
+  }
 
+  protected def registerEventCleanupTimer(
```
[the remark this answers is not in the document]
Comment (fhueske) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/KeyedProcessFunctionWithCleanupState.scala]: We implemented state cleanup as processing time because it is easier to reason about for users and doesn't interfere that much with event-time processing (it is not possible to distinguish timers yet). However, it also has a few short comings such as cleared state when recovering a query from a savepoint (which we don't really encourage at the moment). 

Anyway, introducing event-time state cleanup should definitely go into a separate issue and PR.

[the remark this answers is not in the document]
Comment (liurenjie1024) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/KeyedProcessFunctionWithCleanupState.scala]: The reason why I put it in the same PR is that I don't want it to block this PR, but I also agree that we should move it to a separate one.

**Comment by liurenjie1024** [comment]
@twalthr  I store each column separately to avoid encoding and schema management problems, also to reduce unnecessary data transfer when querying data. Issue multiple requests for multiple fields is one of the drawbacks.
